rfcs/xxxx-authorization/text.md

```diff
@@ -0,0 +1,186 @@
+1. The Explorer needs to know about authz. Currently, it checks the user's identity so it can
+ associate user annotations with the correct user. 
+ But it also needs to apply the authorization
+  logic that's currently in the Portal: only let the owner see private data.
+2. We have a long-lived bearer token (aka obfuscated uuid) authz use case that needs to work in
+ Portal and Explorer.
+
+Right now all the authz information lives in the Portal, so the Explorer has no idea what an "owner" is. 
+Also, the design needs to not harm performance in the explorer: things like caching in the
+instances and the CDN should still work correctly.
+
+
+## [Optional] Product Requirements
+
+### Requirements for obfuscated uuid
+
+From the 0001-data-portal-architecture, under Publishing a Collection Privately:
+
+    Publishing a collection privately runs the full data processing pipeline on the given collection
+    details and datasets and delivers a URL where the submitted may view their generated collection
+    page as well as their dataset(s) in cellxgene. This URL, under the hood, will point to a
+    permanent URL but will be unindexed and obfuscated (though sharable)....
+
+A collection that has visibility=PRIVATE, may have an obfuscated_uuid.  If
+present then the collection can be accessed using the obfuscated_uuid in place of the normal
+collection uuid.  API's like `/dp/v1/collections/{collection_uuid}`, can be accessed through this
+URI: `/dp/v1/collections/{obfuscated_uuid}`, with the following differences:
+ 
+If the obfuscated_uuid is used, then no checks for visibility or user id are performed.  Anyone
+with the obfuscated_uuid will have access.  
+
+The obfuscated_uuid does not expire, however it can be revoked.
+
+An obfuscated_uuid applies to a collection, and not a dataset, and a collection can have only one.
+
+Links to the cellxgene explorer from the portal will need to include some information about the
+obfuscated uuid, so that those possessing the URL can also explore the data.
+For example, instead of `/e/<dataset_name>.cxg`, we use `/e/<obfuscated_uuid>/<dataset_name>.cxg`.
+Or perhaps the obfuscate_uuid can be included as part of the query string.
+ 
+## Detailed Design | Architecture | Implementation
+
+### cellxgene Explorer
+
+#### Handling of private datasets
+
+This task involves checking if the visibility of the dataset is PUBLIC, or if the dataset's
+ owner is the same as the current user_id.  If so, then access to the dataset can be granted.
+Also, this check would need to be done on every API request so streamlining this process may be
+important for performance.
+
+To streamline the process, information about whether the user has access to the dataset can be
+stored in the session cookie.  The session cookie stores information as dictionary.  I propose a
+key named "dataset_access_grant", and a value that is a dictionary with the keys "dataset" and
+"expiration".  The check for the session cookie might look like this.
```

Comment:
Discussion:

I see the updateside of storing dataset access dictionary in the session cookie is to avoid making a machine to machine API to check authorization, which makes response time much faster! 

The only downside I can think of is the cookie size limit of `4096 bytes`. So if a user has many private datasets, it could go over the cookie size limit?

[This article](https://authguidance.com/2017/10/03/api-tokens-claims/) on API AuthZ design seems to combine the best of both worlds (machine to machine authz check + caching). 

From what I understand, it sounds like the architecture is storing a user's dataset access dictionary on a `claims` server. And then a cache layer (key: User's Access Token, value: The user's claim) to provide quick claim lookup

When a user requests for a private dataset access, the API server will send the access token to the `claims` server to find its corresponding `claim`. And then the API server will use the `claim` to check if the user has access to the dataset requested. If so, return the dataset and CloudFront will cache the response with url + cookies as key (I remember our cookies include encoded form of access token?)

The second approach the article talks about uses API Gateway pattern, and talks about using reference token to look up JTW token in the store. And the claims are stored inside a JWT token.

I don't know if our system is set up to be more compatible with option 1 or 2, or if we even need those options 😆 

Just throwing it out there for discussion and learning for myself lol Thank you!

Reply:
I added a long section about caching.  @tihuan thanks for the tip about how Cloudfront works with Cookies.  I based the design on that idea.

Reply:
Awesome thank you!! Glad I could contribute 😆🙏 Noice! So the special cookie gives us a cache size of 1 caching. I think that should be good enough for now and avoid over engineering lol?